src/shared/lua/LuaLib.inl

```diff
@@ -0,0 +1,401 @@
+
+	int mt = lua_gettop( L ); //mt = 1
+	T** ptrHold = ( T** )lua_newuserdata( L, sizeof( T** ) ); //->[2] = empty userdata
+	int ud = lua_gettop( L ); //ud = 2
+
+	if ( ptrHold != NULL )
+	{
+		*ptrHold = obj;
+		lua_pushvalue( L, mt ); // ->[3] = copy of [1]
+		lua_setmetatable( L, -2 ); //[-2 = 2] -> [2]'s metatable = [3]; pop [3]
+		char name[32];
+		tostring( name, sizeof( name ), ptrHold );
+		lua_getfield( L, LUA_REGISTRYINDEX, "DO NOT TRASH" ); //->[3] = value returned from function
+
+		if ( lua_isnil( L, -1 ) ) //if [3] hasn't been created yet, then create it
+		{
+			luaL_newmetatable( L, "DO NOT TRASH" ); //[4] = the new metatable
+			lua_pop( L, 1 ); //pop [4]
+		}
+
+		lua_pop( L, 1 ); //pop [3]
+		lua_getfield( L, LUA_REGISTRYINDEX, "DO NOT TRASH" ); //->[3] = value returned from function
+
+		if ( gc == false ) //if we shouldn't garbage collect it, then put the name in to [3]
+		{
+			lua_pushboolean( L, 1 ); // ->[4] = true
+			lua_setfield( L, -2, name ); //represents t[k] = v, [-2 = 3] = t -> v = [4], k = <ClassName>; pop [4]
+		}
+		else
+		{
+			//In case this is an address that has been pushed
+			//to lua before, we need to set it to nil
+			lua_pushnil( L ); // ->[4] = nil
+			lua_setfield( L, -2, name ); //represents t[k] = v, [-2 = 3] = t -> v = [4], k = <ClassName>; pop [4]
+		}
+
+		if ( IsReferenceCounted<T>() )
+		{
+			// TODO: Remove
+		}
+
+		lua_pop( L, 1 ); // -> pop [3]
+	}
+
+	lua_settop( L, ud ); //[ud = 2] -> remove everything that is above 2, top = [2]
+	lua_replace( L, mt ); //[mt = 1] -> move [2] to pos [1], and pop previous [1]
+	lua_settop( L, mt ); //remove everything above [1]
+	return mt;  // index of userdata containing pointer to T object
+}
+
+
+template<typename T>
+T* LuaLib<T>::check( lua_State* L, int narg )
+{
+	T** ptrHold = static_cast<T**>( lua_touserdata( L, narg ) );
+
+	if ( ptrHold == NULL )
+		return NULL;
+
+	return ( *ptrHold );
+}
+
+
+//private members
+
+template<typename T>
+int LuaLib<T>::thunk( lua_State* L )
+{
+	// stack has userdata, followed by method args
+	T* obj = check( L, 1 ); // get 'self', or if you prefer, 'this'
+	lua_remove( L, 1 ); // remove self so member function args start at index 1
+	// get member function from upvalue
+	RegType* l = static_cast<RegType*>( lua_touserdata( L, lua_upvalueindex( 1 ) ) );
+
+	//at the moment, there isn't a case where NULL is acceptable to be used in the function, so check
+	//for it here, rather than individually for each function
+	if ( obj == NULL )
+	{
+		lua_pushnil( L );
+		return 1;
+	}
+	else
+		return l->func( L, obj ); // call member function
+}
+
+
+
+template<typename T>
+void LuaLib<T>::tostring( char* buff, size_t buff_size, void* obj )
+{
+	snprintf(buff, buff_size, "%p", obj);
+}
+
+
+
+template<typename T>
+int LuaLib<T>::gc_T( lua_State* L )
+{
+	T* obj = check( L, 1 ); //[1] = this userdata
+
+	if ( obj == NULL )
+		return 0;
+
+	if ( IsReferenceCounted<T>() )
```

Comment:
Can we nuke the non-reference counted code path please? In librocket this was used for example for `Event`s which were allocated on the stack. But it's totally unsafe as there is nothing preventing a reference being created to the object which will be dangling after the function containing the stack object returns. This might make sense for librocket as it's hard to get good performance with HTML rendering, but we shouldn't need it.

Reply:
done